
# REST_API_Basics
## Definition, Structure, and Usage through simple examples
Over the course of my career, I’ve come across many technical colleagues who were experts on a given product or technology, but always from a user’s point of view. They knew how to design a solution, sell it to the customer, install it, and customize it to meet the customer’s requirements. And, sometimes, they could even debug rather difficult problems. Yet, when developers asked these technology experts about possible integration of their product with third-party software, or when they were pressed to provide scripts and automate processes, the technology experts were at a loss. They simply did not understand the product from a developer’s perspective. 

And for good reason. There’s a certain amount of magic that happens within an application that is taken care of by the code itself. Many applications you work with today, on PCs, servers, or off in the cloud, include a hidden jewel called the **Application Programming Interface (API)**. In a nutshell, an API makes it possible to interact with an application programmatically. This means that developers can integrate their application with applications in a standardized way. This becomes very useful because it opens up a whole lot of new use cases and business opportunities. But in order for you to take advantage of these opportunities, you need to understand the product’s API well enough so that you can discuss its capabilities with developers.
## Can you REST?
Exploring how you interact programmatically with a product is no easy task. Developers are more familiar with this exercise, but as a non-developer, you might wonder how to get started. The good news is that a de facto standard for building APIs has emerged in the last few years. It’s called **REST (REpresentation State Transfer)**. It was nailed down by Roy Fielding in his PhD thesis. Let me first explain what REST is, and then I’ll define a few important concepts used by RESTFul APIs, such as HTTP and JSON.
## API endpoint
Most of you are familiar with the term HTTP (HyperText Transfer Protocol) and URL (Uniform Resource Locator) because you use a web browser every day. Yet you may not know how they relate to REST and APIs. REST is an architecture that uses a set of principles to describe how networked resources are defined and addressed, leveraging HTTP to build APIs. 

In API land, we use a URL to point to an API and we call it an **endpoint**. You would always start working with an API by knowing what its endpoint is. The following are examples of API endpoints:

* https\://my-oneview-address.mycompany.net/rest
* https\://my-ilo-address.mycompany.net/redfish/v1
* https\://dynamodb.us-west-2.amazonaws.com

## HTTP verbs
The HTTP specification defines how messages are formatted and transmitted, as well as what actions Web servers should take in response to various commands. In HTTP, one uses **verbs** to describe an action that will apply to a given URL or endpoint. When you launch a web browser, the implicit verb is **GET**. However, there are other verbs in the HTTP specifications such as **POST**, **PUT**, **PATCH**, and **DELETE**. REST APIs use these verbs to describe possible actions. The URL in a REST API call is, like in a browser, the target to which the verb must be applied. For example, ` GET http://google.com ` means "retrieve the content of the main index page at google.com, and render it in my browser window". 

Let’s review the most important HTTP verbs:
* GET – Retrieve object instance properties
* PUT/PATCH – Modify object instance properties (two different ways)
* POST – Create a new object instance
* DELETE - Remove an object instance 

## Passing data back and forth
When we know how to use an HTTP verb to describe an action, we can use a URL to point to something in the API, but we need a mechanism to exchange data with the application at the other end. This mechanism is useful to create new items or to retrieve a list of items from the application. We use a couple of techniques to do this.

### HTTP headers
The first technique uses something described by the HTTP specifications called **HTTP headers**. A header allows the attachment of information with the API call as a key/value pair of strings. 

For example, you can add an  `Accept ` header with value ` application/json ` to tell the API to return the result in json. There is no standard for HTTP headers, but there are many well-known headers used by APIs such as **Accept**, **Content-Type**, **X-API-Version**, etc. Headers are useful for passing a simple text value. Think about it as a property for your API call.

### HTTP payload
But what if you want to pass more data to your API? For this, you can use the HTTP message **payload**, also known as the **body** of the API call. This is typically used to specify the credentials to get access to an API or to provide the properties for a new object to create via the API. But to pass data back and forth with an API, you need to make sure the client (calling the API) and the API agree on a format for this data. The payload can be provided in multiple formats such as text, XML or JSON.

### JSON
The most popular format nowadays is called **JSON (for JavaScript Object Notation)**. JSON is a fairly readable machine language which typically looks like a series of Key/Value pairs separated with colons, with string values in quotes and multiple key/values separated with commas. The following is a simple example of a valid JSON snippet:

```json

{
“userName”: “foo@hpe.com”,
“password”: “super-secure-password”
}   
```

JSON is case sensitive, so make sure you have the key names right ("userName" is not the same as "username"). You can express complex values by nesting JSON values and by using square brackets to indicate a collection of similar items as shown below:

```json
{ “members”: [
                { “lastname”: “max”,
                 “firstname”: “mad”,
                 “age”: 50},
                {“lastname”: “james”,
                 “firstname”: “bond”,
                 “age”: 60}],
“count”: 2
}

```

An API might accept more than one input format for payloads, and it may also provide responses in more than one format (JSON, XML, TEXT). Therefore, the convention is to use HTTP headers to specify which format you are going to use to pass data to the API (in your payload) and what data format you expect for the response. 

These headers are: 
* Accept: application/json (please provide me with a response in JSON) 
* Content-Type: application/json (I’m handing out a payload in JSON)

## Response code



## Putting it all together
In summary, an API uses:
* A URL. It also uses the term URI (Uniform Resource Identifier) to specify an object ID or a class of objects
* A verb to specify what action to take on that URL/URI
* One or more HTTP headers to attach certain properties on the API call, such as what format to accept for the response
* A payload to pass data to the API

Let’s look at a few examples using the HPE OneView API:

* GET https\://my-oneview-address.mycompany.net/rest/servers with HEADER Accept set to application/json: Retrieves the list of servers and request results in JSON 
* POST https\://my-oneview-address.mycompany.net/rest/login-sessions with HEADER Content-Type set to application/json and {"userName"="foo","password"="password"} as a JSON payload/body: Creates a new object of type login-session, which is equivalent to logging in the GUI
* DELETE  https\://my-oneview-address.mycompany.net/rest/login-sessions: with HEADER Auth set to session token: Deletes the provided login-session, which is equivalent of logging out in the GUI

## Learning to cURL
Since a browser can only really handle GET, it is clearly not the appropriate tool for manipulating an API. One first tool many people use is a command line tool called cURL. This is available with all Linux distributions and with MacOS. You can open a console and type: ` curl ` to see if you have it installed:

```bash
$ curl
curl: try 'curl --help' or 'curl --manual' for more information
```

If you are running Windows 10, you can get cURL from a PowerShell console (it's actually an alias) or you can download a version of cURL for Windows from https://curl.haxx.se/windows/.


There are many available options but let us focus on the most important ones:

| Option | Meaning                    | Example                                                                         |
|--------|----------------------------|---------------------------------------------------------------------------------|
| H      | HTTP Header                | -H "accept: application/json"                                                   |
|        |                            | -H "content-type: application/json"                                             |
| d      | data or payload            | -d '{"userName":"Administrator","password":"password"}'                         |
| X      | command                    | -X GET https://<ip>/rest/version                                 |
|        |                            | -X POST https://<ip>/rest/login-sessions                         |
| k      | insecure                   | bypasses SSL certificate validation                                             |
| i      | include HEADER in response | useful to check the response status code, but not used when parsing JSON result |
| s      | silent                     | hide progress bar                                                               | 
| x      | use a proxy                | -x http://mycompanyproxy.com:8888                                               |

This is an example of curl call to retrieve HPE OneView API version:

```bash
$ curl -i -k -H "accept: application/json" -X GET https://my-oneviewserver.mycompany.net/rest/version
```

## Postman to the rescue

While cURL is a convenient little tool, it's also a very basic tool. There is another fantastic tool called Postman, which offers a free version that can be used to explore APIs. Visit http://getpostman.com to find the version that works for you (Windows, Mac and Linux). Once installed, start Postman on your machine. You do not need to sign-in to use it for exploring APIs. Since Postman is more than just a client API tool, you’ll want to check out its full feature set. 

## Let’s find a sandbox to give it a try

__Note:__ Make sure you have access to the Internet for this next section.

Millions of APIs are available on the Internet today. Most of them are fee-based and require proper authentication, but a few free APIs are still available, which you can use with or without authentication. I selected a very simple one about a weather report, which doesn’t require authentication to use. You can look up more in this list to find another API, if you prefer: https://github.com/public-apis/public-apis

The API end-point is: https://www.metaweather.com/api/. According to the documentation, we first issue a call to find a location on Earth:

1. Querying API for WOEID of a city, such as Paris

In Postman, make sure the URL is set to https://www.metaweather.com/api/location/search/?query=Paris, and that GET is the selected verb before you hit Send to issue the API call. Ideally you should also add a Header for Accept:application/json to make sure the API returns only JSON. 
You can see in the response section that it is, indeed, provided in JSON, which is somehow readable:

![Postman: GET City of Paris](Pictures/PostmanQuery1.png)

__Note:__ We use cURL instead of Postman to show you live results, but you can do the exercise with Postman on your laptop and get the same experience.

In [1]:
export city="Paris"
curl -s -H "Accept:application/json" -X GET https://www.metaweather.com/api/location/search/?query=$city | jq  "."

[
  {
    "title": "Paris",
    "location_type": "City",
    "woeid": 615702,
    "latt_long": "48.856930,2.341200"
  }
]


From the JSON response, we can extract the woeid (Where On Earth ID) for Paris (615702) and use this value to query for a weather report in Paris: 

In [2]:
woeid=`curl -s -H "Accept:application/json" -X GET https://www.metaweather.com/api/location/search/?query=$city | jq  ".[0].woeid"`
echo "$woeid"

615702


2. Query API for weather report in Paris using WOEID

Still in Postman, make sure the URL is set to https://www.metaweather.com/api/location/615702/ and hit Send again:

![Postman: GET Weather in Paris](Pictures/PostmanQuery2.png)


In [3]:
curl -s -H "Accept:application/json" -X GET https://www.metaweather.com/api/location/$woeid/ | jq "."

{
  "consolidated_weather": [
    {
      "id": 5240644802969600,
      "weather_state_name": "Light Rain",
      "weather_state_abbr": "lr",
      "wind_direction_compass": "SW",
      "created": "2020-06-26T15:37:04.465127Z",
      "applicable_date": "2020-06-26",
      "min_temp": 21.37,
      "max_temp": 30.880000000000003,
      "the_temp": 30.445,
      "wind_speed": 6.32728228284154,
      "wind_direction": 224.3349942837838,
      "air_pressure": 1011.5,
      "humidity": 51,
      "visibility": 14.65752505368647,
      "predictability": 75
    },
    {
      "id": 5443334694764544,
      "weather_state_name": "Light Rain",
      "weather_state_abbr": "lr",
      "wind_direction_compass": "SW",
      "created": "2020-06-26T15:37:07.964039Z",
      "applicable_date": "2020-06-27",
      "min_temp": 17.71,
      "max_temp": 25,
      "the_temp": 24.55,
      "wind_speed": 9.872390210599432,
      "wind_direction": 221.0042766920885,
      "air_pressure": 1011.5,
      "humidity":

As you can see, a fair amount of information is provided by the API, including latitude, longitude, temperature, and weather conditions in Paris (most likely not great :-)). As a developer, you could parse this JSON response and use it in your own application.

Here, we can use jq and its powerful (yet cryptic) syntax to filter out data and see the weather evolution in Paris for the next 5 days.

In [4]:
curl -s -H "Accept:application/json" -X GET https://www.metaweather.com/api/location/$woeid/ | jq ".consolidated_weather[] | { applicable_date, weather_state_name }"

{
  "applicable_date": "2020-06-26",
  "weather_state_name": "Light Rain"
}
{
  "applicable_date": "2020-06-27",
  "weather_state_name": "Light Rain"
}
{
  "applicable_date": "2020-06-28",
  "weather_state_name": "Showers"
}
{
  "applicable_date": "2020-06-29",
  "weather_state_name": "Heavy Cloud"
}
{
  "applicable_date": "2020-06-30",
  "weather_state_name": "Heavy Cloud"
}
{
  "applicable_date": "2020-07-01",
  "weather_state_name": "Showers"
}


Perfect! Optionally, you can redo the sequence for another city on Earth. Give it a try!

## Now what?
Congratulations! You have placed your first API calls using Postman and cURL. This simple API only accepted GET calls, so one might argue everything could have been handled by a standard browser. The advantage of using Postman is that you can save your calls and start building collections of API calls that are useful in your job. We used a very limited API to get the weather report in Paris, but the exact same technique (using Postman) could be applied with other REST APIs, such as the ones from HPE OneView, HPE SimpliVity or Redfish. 


# Time to wrap up
You can now move to the conclusion of this workshop !

* [Conclusion](2-WKSHP-Conclusion.ipynb)